ddl/partition.go

```diff
@@ -515,9 +515,36 @@ func buildTablePartitionInfo(ctx sessionctx.Context, s *ast.PartitionOptions, tb
 			ctx.SetValue(sessionctx.QueryString, newQuery)
 		}
 	}
+
+	partCols, err := getPartitionColSlices(ctx, tbInfo, s)
+	if err != nil || partCols == nil {
+		return errors.Trace(err)
+	}
+
+	for _, index := range tbInfo.Indices {
+		if index.Unique && !checkUniqueKeyIncludePartKey(partCols, index.Columns) {
+			index.Global = config.GetGlobalConfig().EnableGlobalIndex
+		}
+	}
 	return nil
 }
 
+func getPartitionColSlices(sctx sessionctx.Context, tblInfo *model.TableInfo, s *ast.PartitionOptions) (partCols stringSlice, err error) {
+	if s.Expr != nil {
+		extractCols := newPartitionExprChecker(sctx, tblInfo)
+		s.Expr.Accept(extractCols)
+		partColumns, err := extractCols.columns, extractCols.err
+		if err != nil {
+			return nil, err
+		}
+		partCols = columnInfoSlice(partColumns)
+	} else if len(s.ColumnNames) > 0 {
+		partCols = columnNameSlice(s.ColumnNames)
+	}
+	// TODO: Check keys constraints for list, key partition type and so on.
+	return partCols, nil
```

Comment:
Why not return an error here?
Key partition is not yet in master, and I would assume List would also either have a partitioning expression or columns. Hash should have an expression and Key should have a list of column.

Reply:
Just move https://github.com/pingcap/tidb/pull/41437/files#diff-b371d1c1808ba485cc1a00ac338de4bf51b1d4b40fc3cbe274ebfa31580af5a1L3113-L3127 to here.
If we return an error, what is the error message here?

Reply:
I think you can return a custom error, something like an "Internal error, table partition metadata not correct, neither partition expression or list of partition columns".